src/components/SearchBar.tsx

```diff
@@ -0,0 +1,142 @@
+import * as React from 'react';
+import * as _ from 'underscore';
+
+export interface SearchBarProps {
+    onTextChange: Function;
+    onStructuredModeChanged: Function;
+    onDetailedModeChanged: Function;
+    onCategoriesChange: Function;
+    categories: string[];
+}
+
+export interface SearchBarState {
+    expanded: boolean;
+    selectedCategories: string[];
+    structured: boolean;
+    detailed: boolean;
+}
+
+export class SearchBar extends React.Component<SearchBarProps, SearchBarState> {
+
+    constructor(props: SearchBarProps) {
+        super(props);
+        this.state = { expanded: false, selectedCategories: this.props.categories, structured: false, detailed: false };
+    }
+
+    onTextChange(event: any) {
+        this.props.onTextChange(event);
+    }
+
+    onOptionsButtonClick() {
+        this.setState({ expanded: !this.state.expanded });
+    }
+
+    onStructuredModeChange(event: any) {
+        this.props.onStructuredModeChanged();
+        this.setState({ structured: !this.state.structured });
+    }
+
+    onDetailedModeChange(event: any) {
+        this.props.onDetailedModeChanged();
+        this.setState({ detailed: !this.state.detailed });
+    }
+
+    onAllCheckboxChange(event: any) {
+        if (event.target.checked) {
+            this.setState({ selectedCategories: this.props.categories });
+            this.props.onCategoriesChange(this.props.categories);
+        }
+        else {
+            this.setState({ selectedCategories: [] })
+            this.props.onCategoriesChange([]);
+        }
+    }
+
+    onCategoriesChange(event: any) {
+        let categories = this.state.selectedCategories;
+        if (event.target.checked) {
+            if (!_.contains(this.state.selectedCategories, event.target.name)) {
+                categories.push(event.target.name);
+            }
+        }
+        else {
+            categories = _.without(categories, event.target.name);
+        }
+        this.setState({ selectedCategories: categories })
+        this.props.onCategoriesChange(categories);
+    }
+
+    allCategoriesSelected() {
+        return (this.state.selectedCategories.length == this.props.categories.length);
+    }
+
+    isSelectedCategory(category: string) {
+        return _.contains(this.state.selectedCategories, category);
+    }
+
+    setSelectedCategory(event: any) {
+        this.setState({ selectedCategories: [event.target.name] })
+    }
+
+    createCheckbox(name: string, checkboxClassName: string, checked: boolean, onChangeFunc: any, displayText?: string): JSX.Element {
```

Comment:
The new class/interface can contain a method that generates the corresponding `JSX.Element`.